spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/web/jetty/JettyMetricsAutoConfiguration.java

```diff
@@ -0,0 +1,81 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.actuate.autoconfigure.metrics.web.jetty;
+
+
+import java.util.Collections;
+
+import io.micrometer.core.instrument.binder.jetty.JettyStatisticsMetrics;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.handler.HandlerCollection;
+import org.eclipse.jetty.server.handler.StatisticsHandler;
+
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication.Type;
+import org.springframework.boot.web.embedded.jetty.ConfigurableJettyWebServerFactory;
+import org.springframework.boot.web.embedded.jetty.JettyReactiveWebServerFactory;
+import org.springframework.boot.web.embedded.jetty.JettyServletWebServerFactory;
+import org.springframework.boot.web.server.WebServerFactoryCustomizer;
+import org.springframework.context.annotation.Bean;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for {@link JettyStatisticsMetrics}.
+ *
+ * @author Even Holthe
+ * @since 2.1.0
+ */
+@ConditionalOnWebApplication
+@ConditionalOnClass({ JettyStatisticsMetrics.class, Server.class })
+public class JettyMetricsAutoConfiguration {
+	private volatile StatisticsHandler statsHandler;
+
+	@Bean
+	@ConditionalOnMissingBean(JettyStatisticsMetrics.class)
+	public JettyStatisticsMetrics jettyMetrics() {
+		return new JettyStatisticsMetrics(this.statsHandler, Collections.emptyList());
+	}
+
+	@Bean
+	@ConditionalOnWebApplication(type = Type.SERVLET)
+	public WebServerFactoryCustomizer<JettyServletWebServerFactory> serverCapturingServletJettyCustomizer() {
+		return this::statisticsCustomizer;
+	}
+
+	@Bean
+	@ConditionalOnWebApplication(type = Type.REACTIVE)
+	public WebServerFactoryCustomizer<JettyReactiveWebServerFactory> serverCapturingReactiveJettyCustomizer() {
+		return this::statisticsCustomizer;
+	}
+
+	private void wrapExistingHandlersInStatisticsHandler(Server server) {
+		final Handler[] existingHandlers = server.getHandlers();
+		final StatisticsHandler statsHandler = new StatisticsHandler();
+
+		statsHandler.setHandler(new HandlerCollection(existingHandlers));
```

Comment:
There's no need to use a `HandlerCollection` here. A `Server` only ever has a single `Handler` and `getHandlers()` just wraps that single handler in an array.

Reply:
We need to avoid adding a second `StatisticsHandler` to the server. I think it should be possible to get the `Handler` from the server and then walk down through all the wrappers by checking for instances of `HandlerWrapper`.

Reply:
Okay, I've tried to analyze the situation.

1. Get existing `Handler` for the `Server`
2. Check if instance of HandlerWrapper
    1. Get nested child handlers by casting the instance to HandlerWrapper and invoking [`#getChildHandlers()`](https://www.eclipse.org/jetty/javadoc/current/org/eclipse/jetty/server/handler/AbstractHandlerContainer.html#getChildHandlers--) (or maybe [`#getChildHandlersByClass(HandlerWrapper.class)`](https://www.eclipse.org/jetty/javadoc/current/org/eclipse/jetty/server/handler/AbstractHandlerContainer.html#getChildHandlersByClass-java.lang.Class-))
    2. Store all child handlers in a list
    3. Check list of child handlers for any existing instance of `StatisticsHandler`
3. Configure `server.setHandler` and field `statsHandler` to the following:
    - **If no existing `StatisticsHandler`:** Use the existing declared `statsHandler`
    - **If 1 existing `StatisticsHandler`:** Use this instance and remove it from the child handler list
    - **If >1 existing `StatisticsHandler`:** TBD?

Regardless of the outcome of **3**, the `StatisticsHandler#setHandler` must be invoked to wrap the child handlers that was extracted to a list. Here I would guess a `HandlerCollection` would be appropriate?

Reply:
2 is where things get tricky and why we decided not to try to squeeze this into RC2. We basically need to navigate down from the server's handler so that we check to see if any of the wrapped handlers are a `StatisticsHandler`. I don't know exactly how to do this so some experimentation and accompanying tests would be the way to go.

If there happens to be multiple `StatisticHandler`s, I think we can just use the first one that we find. If there's more than one then that's a misconfiguration and I don't think we should add complexity to deal with it.

For 3, if there's no existing `StatisticsHandler` then we need to create one, use it to wrap the server's existing handler, set it on the server, and use it to create the `JettyStatisticsMetrics`. If we've found an existing `StatisticsHandler` then we should leave it in place and use it to create the `JettyStatisticsMetrics`.